### PR TITLE
Add Nintendo Switch support through CMake

### PR DESCRIPTION
**Pull request by AGraber**
This adds support for building librw for the Nintendo Switch through CMake. It uses a CMake executable provided by devkitPro, which uses its own toolchain file to use the correct compilers and paths for includes/libraries, etc.

Since devkitPro is a *de facto* requirement for building homebrew software for the Switch, it is expected that devkirPro is already installed by users that are building for the Switch.

Also includes a GitHub Actions workflow that uses a Docker image provided by devkitPro, for continuous integration.

### CODE REVIEW

**Comment by madebr** [comment]
I think the cmake script needs some more modifications to make it more robust and idiot proof.
Is `NINTENDO_SWITCH` set by the devkitpro nintendo switch cmake toolchain?

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -175,6 +178,15 @@ if(LIBRW_PLATFORM_GL3)
             )
         endif()
     endif()
+
+    if(NINTENDO_SWITCH)
+        target_link_libraries(librw
+            PUBLIC
+                EGL
+                glapi
+                drm_nouveau
```
Comment (madebr) [inline src/CMakeLists.txt:187]: Can we assume these to be exist?
No need for a call to `find_library`?

File: .github/workflows/build-switch.yml [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    types: published
+jobs:
+  build-nintendo-switch:
+    runs-on: ubuntu-latest
+    container: devkitpro/devkita64:latest
+    steps:
+      - uses: actions/checkout@v2
+      - name: "Build files"
+        run: |
+          /opt/devkitpro/portlibs/switch/bin/aarch64-none-elf-cmake -S. -Bbuild .. -DLIBRW_PLATFORM=GL3 -DLIBRW_GL3_GFXLIB=GLFW -DLIBRW_INSTALL=True
```
Comment (madebr) [inline .github/workflows/build-switch.yml:15]: Does it work with the SDL2 backend?

**Comment by madebr** [comment]
Does the imguitest executable inside the github action artifact work?

File: src/CMakeLists.txt (hunk 1 above)
Comment (AGraber) [reply src/CMakeLists.txt:187]: They're always present on the devkitPro Docker image, and can be trivially installed with devkitPro's package manager:
`(dpk-)pacman -S switch-mesa switch-glfw switch-drm_nouveau` (dpk- prefix on Unix systems, not needed on windows). devkitPro's toolchain file makes sure the right include/lib directories are set as well.

A find_library to make sure the libraries are actually there wouldn't hurt I guess.

File: .github/workflows/build-switch.yml (hunk 2 above)
Comment (AGraber) [reply .github/workflows/build-switch.yml:15]: Oh, it's missing indeed. It should work though, I'll add it in a while.

**Comment by AGraber** [comment]
> 
> 
> I think the cmake script needs some more modifications to make it more robust and idiot proof.
> Is `NINTENDO_SWITCH` set by the devkitpro nintendo switch cmake toolchain?
> I'll open a pr on your fork with some suggestions.

Yeah, I agree with that, feel free to let me know any suggestions you have.
Yes, `NINTENDO_SWITCH` is set by their CMake toolchain, as well as a bunch of other variables to make sure it's building with the right compilers and library/include paths.

> 
> 
> Does the imguitest executable inside the github action artifact work?

It should, but it's still in a generic ELF format, the CMake toolchain from devkitPro provides the `nx_generate_nacp` and `nx_create_nro` functions to convert those to a proper .nro that can be loaded by the hbmenu. I'll add those to the CMakeFile soon as well.

File: src/CMakeLists.txt [hunk 3]
```diff
@@ -175,6 +178,19 @@ if(LIBRW_PLATFORM_GL3)
             )
         endif()
     endif()
+
+    if(NINTENDO_SWITCH)
+        find_library(SWITCH_EGL_LIBRARY EGL)
```
Comment (madebr) [inline src/CMakeLists.txt:183]: I forgot about this pr.
I will still create a pr on your fork with various changes to make the generated tarball relocatable (=movable between systems).

**Comment by madebr** [comment]
I guess the GL3 platform is the only viable platform for Nintendo Switch?
Does a null backend make sense?

**Comment by AGraber** [comment]
> 
> 
> I guess the GL3 platform is the only viable platform for Nintendo Switch?
> Does a null backend make sense?

A null backend should work in theory, but I honestly see little point. Even from a development perspective, a null backend for debugging purposes isn't viable since there's no proper debugging tools for the Switch yet.

I recently tested the imguitest executable. Unfortunately, current GitHub Actions artifacts crash because there's no registered glfw monitors, (in theory there should be at least one, the Switch's screen). I made a PR to fix that at the glfw repo form devkitPro (devkitPro/glfw#3), and with those changes it works fine.

It may be hard to tell, but this is actual footage from the Switch - mouse input is emulated through touch, implemented by glfw itself:


https://user-images.githubusercontent.com/18301034/122422523-82661a80-cf8d-11eb-9f54-d4de204c4c07.mp4

**Comment by madebr** [comment]
@AGraber
Thanks for the answer and the video. Great to see it working.

I'm going to leave the null renderer for now.
How we set the default renderer + handle playstation should be handled different, I think.

Another question:
the binary tarball contains 2 binaries for each installed binary: e.g. `imguitest` and `imguitest.nro`.
Would it be ok to only install the `.nro`?

**Comment by AGraber** [comment]
The ELF binaries might be useful when looking into crashes from Switch builds because it contains debug info and you can run tools like addr2line, etc. You can reverse the NRO conversion but I think some data gets lost and there's no good tools for that either.On 18 Jun 2021 00:41, Anonymous Maarten ***@***.***> wrote:
@AGraber
Thanks for the answer and the video. Great to see it working.
I'm going to leave the null renderer for now.
How we set the default renderer + handle playstation should be handled different, I think.
Another question:
the binary tarball contains 2 binaries for each installed binary: e.g. imguitest and imguitest.nro.
Would it be ok to only install the .nro?

**Comment by madebr** [comment]
@AGraber 
I've opened https://github.com/AGraber/librw/pull/1

**Review by madebr — APPROVED** [review]

**Pull request merged**